Samples/src/OracleEmulation/Emulator.qs

```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+
+namespace Microsoft.Quantum.Extensions.Emulation
+{
+    /// # Summary
+    /// Emulate the effect of a classical oracle by permuting the basis states of the simulator's
+    /// wavefunction such that
+    ///     $\ket{x}\ket{y}\ket{w} \rightarrow \ket{x}\ket{f(x, y)}\ket{w}$,
+    /// with registers x, y, w and the oracle function f. 
+    ///
+    /// # Input
+    /// ## oracle
+    /// A function that defines the action of the oracle on the computational
+    /// basis states of the two registers x, y. The mapping
+    ///		$(x, y) \rightarrow (x, z=f(x, y))$
+    /// must be a bijective mapping on the basis states.
+    /// ## xbits
+    /// Input register x.
+    /// ## ybits
+    /// Output register y.
+    operation EmulateOracle(oracle : ((Int, Int) -> Int), xbits : Qubit[], ybits : Qubit[]) : Unit
```

Comment:
Thanks for the detailed review, Chris! I'll address the more stylistic comments later and will first reply to the general design points.

Regarding the point raised here, I agree `EmulateOracle` is essentially an intrinsic operation specific to one target machine. And, while there may be strong arguments against machine-specific constructs in Q# _on the language level_, I'm not sure we should rule out optimized Q# operations based on machine-specific primitives/intrinsics. In this case it is a simulator backend, but in the future it might be hardware capable of, say, a special multi-qubit operation that I'd like to use directly in my algorithm. Pretending that any piece of Q# code may be run on any target machine would ignore the fact that we'll have to think about the strengths and limitations of our target machine for quite some time.

NB: In principle, any call to `EmulateOracle` could be automatically mapped to a generic target machine by running a synthesis algorithm (as @msoeken is alluding to), but that's not necessarily what you want to do in practice.

Reply:
Tagging @alan-geller @bettinaheim @thomashaener @vadym-kl @matthiastroyer, who might have thoughts on machine-specific intrinsics. I realize this might establish some precedence, even if this is only in a sample for the moment.

Reply:
Discussing with @vadym-kl, it might be nice to have something like a `PermutationOracle` in `Canon` that implements the current interface of `EmulateOracle` via a synthesis algorithm. That would bring similar rapid prototyping benefits to generic target machines (with cost exponential in the register sizes of course); and it would most clearly demonstrate that a call to `EmulateOracle` could be implemented on any target machine.

Of course, more advanced synthesis algorithms again raise similar questions concerning choice of an optimal backend implementation depending on target machine capabilities.

Reply:
In offline discussion with @cgranade, @alan-geller, @martinquantum, @cpalmer2020, we agreed that (please comment if I misremember):

1. The operation should be named according to its effect rather than its machine-specific implementation, and hence `PermutationOracle` is more appropriate.
2. A generic, machine-agnostic implementation of `PermutationOracle` would be very nice, but is not a precondition for accepting the machine-specific (emulated) implementation. In general, the minimum requirement for accepting a machine-specific operation is that a generic version would undoubtedly be implementable. In other words, operations with unphysical effects are not acceptable, but it is clear that permutation oracles can be implemented for any machine (see synthesis methods as mentioned before).
3. As long as there is no generic implementation, the operation should not be `intrinsic`, but rather have a body with a `fail` statement. This makes it more apparent where an implementation must be provided when executing on a target machine that does not provide its own implementation.

The corresponding changes have been pushed.